crates/bevy_ecs/src/schedule/auto_insert_apply_deferred.rs

```diff
@@ -0,0 +1,164 @@
+use std::collections::BTreeSet;
+
+use bevy_utils::{
+    petgraph::{
+        graphmap::GraphMap,
+        Directed,
+        Direction::{Incoming, Outgoing},
+    },
+    HashMap,
+};
+
+use crate::system::IntoSystem;
+
+use super::{
+    apply_deferred, is_apply_deferred, NodeId, ReportCycles, ScheduleBuildError, ScheduleBuildPass,
+    ScheduleGraph, SystemNode,
+};
+
+/// A [`ScheduleBuildPass`] that inserts [`apply_deferred`] systems into the schedule graph
+/// when there are [`Deferred`](crate::prelude::Deferred)
+/// in one system and there are ordering dependencies on that system. [`Commands`](crate::system::Commands) is one
+/// such deferred buffer.
+///
+/// This pass is typically automatically added to the schedule. You can disable this by setting
+/// [`ScheduleBuildSettings::auto_insert_apply_deferred`](crate::schedule::ScheduleBuildSettings::auto_insert_apply_deferred)
+/// to `false`. You may want to disable this if you only want to sync deferred params at the end of the schedule,
+/// or want to manually insert all your sync points.
+#[derive(Debug, Default)]
+pub struct AutoInsertApplyDeferredPass {
+    no_sync_edges: BTreeSet<(NodeId, NodeId)>,
+    auto_sync_node_ids: HashMap<u32, NodeId>,
+}
+
+/// If added to a dependency edge, the edge will not be considered for auto sync point insertions.
```

Comment:
This should have a doc test or link to a type that has a doc test :)

Reply:
We only add this inside `after_ignore_deferred` and `before_ignore_deferred`. I intend to remove these two APIs and replace them with the API I described above. I can certainly add a doc test in that PR.